Implement Ursula test for Swift

yangzhang-ibm opened the pull request:
Implement Ursula test with Swift enabled (based on current heat provisioner approach for ursula PR tests and ceph tests). Once the swift deployment and tests are enabled, it can be included in the current nightly ursula test jenkins job which already has ceph enabled since they are sharing the same set of ursula env "envs/example/ci-ceph_swift".

The test cases for swift in this PR are only very basic ones to make sure swift is deployed properly and working well. After the Openstack is deployed with swift, glance will be configured with swift as backend store by default, so that the swift component can also be verified by the general image creation and query actions during the existing ursula tests.

bbc-jenkins commented:
Can one of the admins verify this patch?

*  \"ok to test\" to accept this pull request for testing
*  \"test this please\" for a one time test run
*  \"add to whitelist\" to add the author to the whitelist

yangzhang-ibm commented:
@j2sol @timrchavez @notnownikki Please help review this PR, thanks!

timrchavez commented:
ok to test

timrchavez commented on envs/example/ci-ceph_swift/group_vars/all.yml:
```diff
@@ -3,6 +3,16 @@ stack_env: example_ci-ceph_swift
 state_path_base: /opt/stack/data
 
+swift_floating_ip: "{{ hostvars[groups['swiftnode'][0]][primary_interface]['ipv4']['address'] }}"
+swift_undercloud_floating_ip: "{{ swift_floating_ip }}"
```
Not seeing where else you use `swift_floating_ip` can we collapse these two variables into one?

yangzhang-ibm replied:
The only place uses `swift_undercloud_floating_ip` is the code of https://github.com/blueboxgroup/ursula/blob/master/roles/swift-common/templates/check-swift-dispersion.sh as follows, it seems ok to remove that definition of `swift_undercloud_floating_ip` and use `swift_floating_ip` only. I will test that.
```
#!/bin/bash
#
# check swift dispersion
# only run on node with the floating ip

if ifconfig | grep {{ swift_undercloud_floating_ip | default( swift_floating_ip ) }} > /dev/null 2>&1; then
  sudo /etc/sensu/plugins/check-swift-dispersion.py -c 98 -d 99 -o 98 -n 99
fi;
```

timrchavez commented on envs/example/ci-ceph_swift/group_vars/all.yml:
```diff
@@ -3,6 +3,16 @@ stack_env: example_ci-ceph_swift
+  - name: "{{ fqdn }}"
+    ip: "{{ undercloud_floating_ip }}"
+  - name: "{{ swift_fqdn }}"
+    ip: "{{ swift_undercloud_floating_ip }}"
```
This will override what's in defaults-2.0.yml which may be okay, but what's provided by https://github.blueboxgrid.com/blue-box-cloud/sitecontroller/blob/master/roles/jenkins-slave/files/scripts/generate_logging_extra_vars.sh#L13-L17 will override it :/ -- we really need to come up with a better way to handle etc entries -- I think for logging I"ll need just use `lineinfile`

yangzhang-ibm replied:
For /etc/hosts on our Openstack nodes, it can only be written by https://github.com/blueboxgroup/ursula/blob/master/roles/common/tasks/networking.yml which is using a template:
```
- name: hosts file
  template: src=etc/hosts dest=/etc/hosts owner=root group=root mode=0644
            backup=yes
  when: etc_hosts is defined
  tags: etc_hosts
```

```
{% for entry in etc_hosts -%}
{{ entry.ip }} {{ entry.name }}
{% endfor %}
```

Btw, it seems the logging part you mentioned doesn't have any impact to our ursula test, right?

notnownikki replied:
As far as I can see, this is fine for ursula because the fqdn entry that it would have overwritten is duplicated here.

timrchavez replied:
Logging is enabled for that nightly log and we do `test/run deploy --extra-vars "@${WORKSPACE}/logging_extra_vars.yml"` which will overwrite your etc_host with the etc_host we define for logging.  This mean that anything you define in etc_host that's not in whatever the logging stuff defines, will be lost.  I'll work on a patch to use lineinfile for logging to eliminate this problem.

timrchavez commented on envs/example/ci-ceph_swift/group_vars/all.yml:
```diff
@@ -20,3 +30,41 @@ cinder:
+  enabled: True
+  allow_versions: True
+  disks:
+    - disk: vdb
```
This could just be a list.... e.g.

```
disks:
  - vdb
  - foo
  - bar
```

yangzhang-ibm replied:
That style with a simple list worked before, but it will cause an error with latest code of swift roles in ursula. So the way I am using is just to fit the latest ursula code. : ) 
Btw, you can also see that format here which was updated 13 days ago: https://github.com/blueboxgroup/ursula/blob/master/envs/example/swift/group_vars/all.yml

timrchavez replied:
:+1:

timrchavez commented on playbooks/ci-ceph_swift/tasks/pre-deploy.yml:
```diff
@@ -35,6 +35,32 @@
+  hosts: swiftnode
+  tasks:
+    - name: umount ephemeral disk on swiftnode
+      shell: umount /dev/{{ item.disk }}
```
See above comment... no need to use dictionaries for things that can be represented by strings.

yangzhang-ibm replied:
As mentioned above, it's to fit latest swift roles in ursula code.

timrchavez replied:
:+1:

timrchavez commented on playbooks/ci-ceph_swift/tasks/pre-deploy.yml:
```diff
@@ -35,6 +35,32 @@
+    - name: deploy fake lshw file /usr/bin/lshw
+      template: src=../templates/usr/bin/lshw
+                dest=/usr/bin/lshw
+                owner=root group=root mode=0755
```
I really wish there was a better way of doing this... you sure update-alternatives does not work??? :P

yangzhang-ibm replied:
I assume that update-alternatives should work (maybe I had some misunderstanding about how it works in the very beginning), but anyway, the benefit of my above approach is that it can minimize the impact of replacing the original lshw script by wrapping the output of original lshw in my fake one if needed (i.e. when it is not running `lshw -C disk`).

yangzhang-ibm replied:
Anyway, we should always do this update-alternatives stuff in our pre-deploy.yml, it can't be included in the standard list of update-alternatives for swift roles in ursula code since it is valid for our test in VMs only.

yangzhang-ibm replied:
After a testing with update-alternatives, I found that it is not suitable here, detailed reason is:
```
ubuntu@ceph-swift-test-swift-0:~$ sudo cat /etc/sudoers.d/swiftops
swiftops ALL= NOPASSWD: /usr/local/bin/swifttool
swiftops ALL= NOPASSWD: /usr/bin/lshw
swiftops ALL= NOPASSWD: /usr/local/bin/swift-init
```
Our alternatives for lshw /usr/local/bin/lshw is not registered in above file which defines the sudoer permission of swiftops user, so password was required during running swifttool with user swiftops.
The solution is to use my previous approach to rename original /usr/bin/lshw to /usr/bin/lshw.original and wrap it in my fake one /usr/bin/lshw.

timrchavez commented on playbooks/ci-ceph_swift/templates/usr/bin/lshw:
```diff
@@ -0,0 +1,23 @@
+  echo "  *-disk:$i"
+  echo "       logical name: /dev/${arr_label[$i]}"
+  echo "       serial: ${hostname}-${arr_label[$i]}"
+  echo "       size: ${arr_size[$i]}GiB"
```
I tend to like using bash heredoc's for this sort of thing... e.g.
```
cat << EOF
  *-disk:$i
      logical name: /dev/${arr_label[$i]}"
      serial: ${hostname}-${arr_label[$i]}"
      size: ${arr_size[$i]}GiB
EOF
```

yangzhang-ibm replied:
Yeah, very good point, let me refine that part. Thanks!

timrchavez commented:
Definitely want @j2sol to look at this.  Also, we should identify whoever the storage squad lead is and have them look too.  They might be interested in this work too :)

timrchavez commented:
"ok to test"

notnownikki commented on playbooks/ci-ceph_swift/tasks/pre-deploy.yml:
```diff
@@ -35,6 +35,32 @@
+
+- name: deploy fake lshw file on swiftnode
+  hosts: swiftnode
+  tasks:
```
Just want to make sure of something here. This playbook is only ever going to be run once, to prepare nodes for deployment, and then never run again, is that correct?

notnownikki replied:
I mean it totally looks that way, I just wanted to make sure of that :)

yangzhang-ibm replied:
Yes, that's correct.

timrchavez commented on playbooks/ci-ceph_swift/templates/usr/bin/lshw:
```diff
@@ -0,0 +1,25 @@
+
+# Output additional content for disks
+arr_label=(`lsblk -db |awk '{if(NR!=1){print $1}}'`)
+arr_size=(`lsblk -db |awk '{if(NR!=1){print $4/1024/1024/1024}}'`)
```
1. No need for the conditional... `NR>1 {print}`
2. 1024/1024/1024 should just be a variable defined as the size of 1GiB for clarity... e.g. 
```
GB=1073741824
lsblk -db |awk -v gb=$GB 'NR>1{print $4/gb}'
```

timrchavez commented on test/setup:
```diff
@@ -80,6 +80,33 @@ else
 
 echo "controller_primary: ${CONTROLLER_IP}" >> ${ROOT}/envs/test/group_vars/all.yml
+echo "ursula_env_path: ${ROOT}/envs/test" >> ${ROOT}/envs/test/group_vars/all.yml
+
```
More that one entry... we should probably convert into a heredoc....

timrchavez commented on playbooks/ci-ceph_swift/tasks/pre-deploy.yml:
```diff
@@ -35,6 +35,34 @@
+    - name: install lshw alternatives
+      alternatives: name=lshw_ursula_test
+                    path=/usr/local/bin/lshw_ursula_test
+                    link=/usr/local/bin/lshw
```
Woo

yangzhang-ibm commented:
Hi @timrchavez, I have updated my code changes as your suggestions, please review again, thanks!
New changes were done to the following files:
```
modified:   test/setup
renamed:   playbooks/ci-ceph_swift/templates/:   usr/bin/lshw -> usr/local/bin/lshw_ursula_test
modified:   playbooks/ci-ceph_swift/tasks/pre-deploy.yml
```

timrchavez commented:
:+1:

timrchavez commented:
ok to test

timrchavez commented:
ok to test

yangzhang-ibm commented:
@timrchavez @j2sol The Ursula PR test #248 for this was finished successfully. Further verification can be done by nightly Ursula test with ceph & swift enabled after this PR gets merged. Thanks!

j2sol commented on playbooks/ci-ceph_swift/tasks/pre-deploy.yml:
```diff
@@ -35,6 +35,32 @@
     - name: umount ephemeral disk on ceph_osds
       shell: umount /dev/{{ item }}
       with_items: "{{ ceph.disks }}"
+      ignore_errors: true
```
This should be `failed_when: False` instead of `ignore_errors: true`.

yangzhang-ibm replied:
Okay, thanks for your suggestion, let me change lines like that.

j2sol commented on playbooks/ci-ceph_swift/tasks/pre-deploy.yml:
```diff
@@ -35,6 +35,32 @@
+    - name: umount ephemeral disk on swiftnode
+      shell: umount /dev/{{ item.disk }}
+      with_items: "{{ swift.disks }}"
+      ignore_errors: true
```
`failed_when: False`

j2sol commented on playbooks/ci-ceph_swift/tasks/pre-deploy.yml:
```diff
@@ -35,6 +35,32 @@
+  tasks:
+    - name: rename the original /usr/bin/lshw
+      command: mv /usr/bin/lshw /usr/bin/lshw.original
+    - name: deploy fake lshw file /usr/bin/lshw
```
Please put a blank line between tasks.

yangzhang-ibm replied:
Added blank lines. Thanks!

j2sol commented on playbooks/ci-ceph_swift/tasks/pre-deploy.yml:
```diff
@@ -35,6 +35,32 @@
+  hosts: swiftnode
+  tasks:
+    - name: rename the original /usr/bin/lshw
+      command: mv /usr/bin/lshw /usr/bin/lshw.original
```
Instead of moving the original, just use the `backup` option to the `template` module and it'll create a timestamped backup file of the file being replaced.

j2sol commented on playbooks/ci-ceph_swift/templates/usr/bin/lshw:
```diff
@@ -0,0 +1,26 @@
+# run the original lshw command
+if [ "$1" != "-C" -o "$2" != "disk" ]; then
+  if [ -f /usr/bin/lshw.original ]; then
+    sudo /usr/bin/lshw.original
```
Would this even work? That path isn't in the sudoers configuration. But isn't this script itself called with sudo? Would you need to re-sudo here?

j2sol replied:
Also, if we're going to use the original lshw, shouldn't we exit at this point, so we don't add extra data?

yangzhang-ibm replied:
Yeah, these are very good points! I have refined that part as follows which with your suggestions applied and also passing all parameters to the original lshw script to avoid any gap in the final output:
```
# run the original lshw command
if [ "$1" != "-C" -o "$2" != "disk" ]; then
  if [ -f /usr/bin/lshw.original ]; then
    /usr/bin/lshw.original $@
    exit 0
  fi
fi
```

yangzhang-ibm replied:
@j2sol Please let me know which one in above 3 options do you prefer, thanks!

j2sol commented on playbooks/tests/tasks/swift.yml:
```diff
@@ -0,0 +1,10 @@
+  tasks:
+  - name: Openstack service list should include swift
+    shell: . /root/stackrc; openstack service list | grep swift
+  - name: Openstack endpoint list should include swift
```
Please put a blank line between tasks.

yangzhang-ibm replied:
Added blank lines. Thanks!

j2sol commented on test/gen-cert:
```diff
@@ -63,7 +63,12 @@ cat >> req.conf <<eof
 basicConstraints = CA:FALSE
 keyUsage = nonRepudiation, digitalSignature, keyEncipherment, keyCertSign
-subjectAltName = DNS.1:${DOMAIN}
+#subjectAltName = DNS.1:${DOMAIN}
```
Why has this been commented out?

yangzhang-ibm replied:
Since the DOMAIN passed from `test/setup` script was `openstack.example.com`, I was trying to use multiple entries for the Subject Alternative Names to enable swift fqdn as follows:
```
subjectAltName = @alt_names
[alt_names]
DNS.1 = ${DOMAIN}
DNS.2 = swift.${DOMAIN}
```

But I thought it again and want to use a better approach:
1. Pass `*.openstack.example.com` as DOMAIN to `gen-cert`.
2. Remove "*." from above DOMAIN as a FILE_PREFIX, and use it instead of original DOMAIN for the name of temp files for cert generating.

yangzhang-ibm replied:
Unfortunately, I encountered the following issue with ssl cert with "*openstack.example.com" although it works well for other services on controller:
```
# swift list
hostname 'swift.openstack.example.com' doesn't match '*openstack.example.com'
```

So I have to revert this part to use my previous approach with 2 lines for Subject Alternative Names in the ssl cert generated by `gen-cert` script:
```
subjectAltName = @alt_names

[alt_names]
DNS.1 = ${DOMAIN}
DNS.2 = swift.${DOMAIN}
```

yangzhang-ibm commented:
@j2sol I have done all the refinements as your above comments mentioned, you can expand your original comments to see my explanations. Please help review again especially for the changes in following 3 files:
``` 
playbooks/ci-ceph_swift/templates/usr/bin/lshw
test/gen-cert
test/setup
```

yangzhang-ibm commented:
The latest Ursula PR test (# 253) after my updates to the commit was finished successfully.
